src/provider/mercury.rs

```diff
@@ -0,0 +1,1427 @@
+      rhs += rhs;
+      rhs += zeta * arg.s_zeta + zeta_inv * arg.s_zeta_inv;
+
+      assert_eq!(lhs, rhs);
+
+      if lhs != rhs {
+        return Err(NovaError::ProofVerifyError {
+          reason: "IPA check failed".to_string(),
+        });
+      }
+    }
+
+    let g1 = Commitment::new(DlogGroup::group(&vk.G));
+    let g2 = <<E as Engine>::GE as PairingGroup>::G2::gen();
+    let tau2 = <E::GE as PairingGroup>::G2::group(&vk.tau_H);
+    let lr = g2;
+    let rr = tau2;
+
+    // * Check f(X) / (X^b - alpha) = (q(X), g(x))
+    // * 5. Adapted from Mercury Section 6. Step 4. (f)
+    // The pairing check in Step 4. (f) is
+    // Pairing 1 LHS = comm_f - [zeta^b - alpha] comm_q - [g(zeta)]_1
+    // Pairing 1 RHS = [1]_2
+    // Pairing 2 LHS = comm_quot_f
+    // Pairing 2 RHS = [tau - zeta]_2
+    //
+    // It is adapted into the following pairing checks:
+    // Pairing 1 LHS: comm_f + [zeta^b - alpha] comm_q - [g(zeta)]_1 + [zeta] * comm_quot_f
+    // Pairing 1 RHS: g1
+    // Pairing 2 LHS: comm_quot_f
+    // Pairing 2 RHS: [tau]_2
+    //
+    // The verifier's checks are batched as follows:
+    // Denote the above as LHS_1, RHS_1=g1, LHS_2, RHS_2=[tau]_2.
+    // Denote the KZG pairing of BDFH20 (Step 4. (g)) as LHS_1', RHS_1'=g1, LHS_2', RHS_2'=[tau]_2.
+    // The verifier samples a random d and checks
+    // e(LHS_1 + LHS_1' * d, g1) = e(LHS_2 + LHS_2' * d, [tau]_2)
+    let (lhs_1_1, lhs_2_1) = {
+      let zeta_b = zeta_b_one * zeta;
+      let zeta_b_alpha = zeta_b - alpha;
+
+      let scalars = [-zeta_b_alpha, -arg.g_zeta, zeta];
+
+      let bases = [arg.comm_q, g1, arg.comm_quot_f]
+        .into_iter()
+        .map(|b| b.into_inner().affine())
+        .collect::<Vec<_>>();
+
+      // * Main Cost (Verifier) I: MSM of 3
+      let ll = *comm_f + Commitment::new(E::GE::vartime_multiscalar_mul(&scalars, &bases));
+      let rl = arg.comm_quot_f;
+
+      #[cfg(debug_assertions)]
+      {
+        let pairing_l = E::GE::pairing(&ll.into_inner(), &lr);
+        let pairing_r = E::GE::pairing(&rl.into_inner(), &rr);
+
+        assert!(pairing_l == pairing_r);
+      }
+
+      (ll, rl)
+    };
+
+    // * Check KZG
+    // * 6. Mercury Section 6. Step 4. (g)
+    let (lhs_1_2, lhs_2_2) = extract_pairing_to_verify_batch_evaluation::<E>(
+      BatchEvaluations {
+        zeta,
+        zeta_inv,
+        alpha,
+        g_zeta: arg.g_zeta,
+        g_zeta_inv: arg.g_zeta_inv,
+        h_zeta: arg.h_zeta,
+        h_zeta_inv: arg.h_zeta_inv,
+        h_alpha,
+        s_zeta: arg.s_zeta,
+        s_zeta_inv: arg.s_zeta_inv,
+        d_zeta,
+      },
+      BatchEvaluationCommitments::<E> {
+        comm_g: arg.comm_g,
+        comm_h: arg.comm_h,
+        comm_s: arg.comm_s,
+        comm_d: arg.comm_d,
+        comm_w: arg.comm_w,
+        comm_w_prime: arg.comm_w_prime,
+      },
+      g1,
+      transcript,
+    )?;
+
+    // Check Pairing of 3. and 4.
+    let d = E::Scalar::random(rand_core::OsRng);
```

Comment:
I think we want to obtain this challenge from the transcript. The main reason is otherwise an on-chain verifier cannot do the same.

Reply:
A side effect of obtaining this challenge from the transcript is that the prover has to squeeze a challenge as well to keep the transcript in the same state.